Sources/LeafKit/LeafTag.swift

```diff
@@ -0,0 +1,40 @@
+public protocol LeafTag {
+    func render(_ ctx: LeafContext) throws -> LeafData
+}
+
+public var defaultTags: [String: LeafTag] = [
+    "lowercased": Lowercased(),
+    "index": LoopIndex(),
+    "isFirst": LoopIsFirst(),
+    "isLast": LoopIsLast()
+]
+
+struct Lowercased: LeafTag {
+    func render(_ ctx: LeafContext) throws -> LeafData {
+        guard let str = ctx.parameters.first?.string else {
+            throw "unable to lowercase unexpected data"
+        }
+        return .init(.string(str.lowercased()))
+    }
+}
+
+struct LoopIndex: LeafTag {
+    func render(_ ctx: LeafContext) throws -> LeafData {
+        guard let index = ctx.data["index"] else { throw "Loop accessor called on non-loop" }
```

Comment:
Also, may I suggest a more formal phrasing along the lines of `"Illegal use of #(index) outside #for loop"`, and similarly for the other tags (the phrasing is not critical, what's most important IMO is that the error specify which tag was incorrectly used, especially given the prevailing absence of source location info).

Reply:
Agreed